crates/services/txpool_v2/src/verifications.rs

```diff
@@ -0,0 +1,232 @@
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct InputComputationVerifiedTx(Checked<Transaction>);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct FullyVerifiedTx(Checked<Transaction>);
+
+pub enum TransactionVerifState {
+    Unverified(UnverifiedTx),
+    BasicVerified(BasicVerifiedTx),
+    InputDependenciesVerified(InputDependenciesVerifiedTx),
+    InputComputationVerified(InputComputationVerifiedTx),
+    FullyVerified(FullyVerifiedTx),
+}
+
+impl UnverifiedTx {
+    pub async fn perform_basic_verifications(
+        self,
+        current_height: BlockHeight,
+        consensus_params: &ConsensusParameters,
+        gas_price_provider: &impl GasPriceProvider,
+    ) -> Result<BasicVerifiedTx, Error> {
+        if self.0.is_mint() {
+            return Err(Error::NotSupportedTransactionType);
+        }
+        let tx = self
+            .0
+            .into_checked_basic(current_height, consensus_params)?;
+        let gas_price = gas_price_provider.next_gas_price().await?;
+        let tx = verify_tx_min_gas_price(tx, consensus_params, gas_price)?;
+        Ok(BasicVerifiedTx(tx))
+    }
+}
+
+impl BasicVerifiedTx {
+    pub fn perform_inputs_verifications<PSView, PSProvider>(
+        self,
+        pool: TxPool<PSProvider>,
+        version: ConsensusParametersVersion,
+    ) -> Result<InputDependenciesVerifiedTx, Error>
+    where
+        PSProvider: AtomicView<LatestView = PSView>,
+        PSView: TxPoolPersistentStorage,
+    {
+        let pool_tx = checked_tx_into_pool(self.0, version)?;
+
+        let transaction = pool
+            .read()
+            .can_insert_transaction(pool_tx)?
+            .into_transaction();
+        let checked_transaction: CheckedTransaction = transaction.into();
+        Ok(InputDependenciesVerifiedTx(checked_transaction.into()))
+    }
+}
+
+impl InputDependenciesVerifiedTx {
+    pub fn perform_input_computation_verifications<M>(
+        self,
+        consensus_params: &ConsensusParameters,
+        wasm_checker: &impl WasmChecker,
+        memory: M,
+    ) -> Result<InputComputationVerifiedTx, Error>
+    where
+        M: Memory + Send + Sync + 'static,
+    {
+        if let Transaction::Upgrade(upgrade) = self.0.transaction() {
+            if let UpgradePurpose::StateTransition { root } = upgrade.upgrade_purpose() {
+                wasm_checker
+                    .validate_uploaded_wasm(root)
+                    .map_err(Error::WasmValidity)?;
+            }
+        }
+
+        let parameters = CheckPredicateParams::from(consensus_params);
+        let tx = self.0.check_predicates(&parameters, memory)?;
+
+        Ok(InputComputationVerifiedTx(tx))
+    }
+}
+
+impl InputComputationVerifiedTx {
+    pub fn perform_final_verifications(
+        self,
+        consensus_params: &ConsensusParameters,
+    ) -> Result<FullyVerifiedTx, Error> {
+        let tx = self.0.check_signatures(&consensus_params.chain_id())?;
```

Comment:
It is better to verify the signatures before predicates since it is cheaper

Reply:
Done